mbuild/packing.py

```diff
@@ -176,27 +175,31 @@ def fill_box(compound, n_compounds=None, box=None, density=None, overlap=0.2,
     box_maxs -= edge * 10
 
     # Build the input file for each compound and call packmol.
-    filled_pdb = tempfile.mkstemp(suffix='.pdb')[1]
-    input_text = PACKMOL_HEADER.format(overlap, filled_pdb, seed)
-
-    for comp, m_compounds, rotate in zip(compound, n_compounds, fix_orientation):
-        m_compounds = int(m_compounds)
-        compound_pdb = tempfile.mkstemp(suffix='.pdb')[1]
-        comp.save(compound_pdb, overwrite=True)
-        input_text += PACKMOL_BOX.format(compound_pdb, m_compounds,
-                           box_mins[0], box_mins[1], box_mins[2],
-                           box_maxs[0], box_maxs[1], box_maxs[2],
-                           PACKMOL_CONSTRAIN if rotate else "")
-
-    _run_packmol(input_text, filled_pdb, temp_file)
-
-    # Create the topology and update the coordinates.
-    filled = Compound()
-    for comp, m_compounds in zip(compound, n_compounds):
-        for _ in range(m_compounds):
-            filled.add(clone(comp))
-    filled.update_coordinates(filled_pdb)
-    filled.periodicity = np.asarray(box.lengths, dtype=np.float32)
+    filled_pdb = _new_pdb_file()
+    try:
+        input_text = PACKMOL_HEADER.format(overlap, filled_pdb.name, seed)
+        for comp, m_compounds, rotate in zip(compound, n_compounds, fix_orientation):
+            m_compounds = int(m_compounds)
+            compound_pdb = _new_pdb_file()
+            comp.save(compound_pdb.name, overwrite=True)
+            input_text += PACKMOL_BOX.format(compound_pdb.name, m_compounds,
+                               box_mins[0], box_mins[1], box_mins[2],
+                               box_maxs[0], box_maxs[1], box_maxs[2],
+                               PACKMOL_CONSTRAIN if rotate else "")
+
+        _run_packmol(input_text, filled_pdb, temp_file)
+
+        # Create the topology and update the coordinates.
+        filled = Compound()
+        filled = _create_topology(filled, compound, n_compounds)
+        filled.update_coordinates(filled_pdb.name)
+        filled.periodicity = np.asarray(box.lengths, dtype=np.float32)
+
+    finally:
+        filled_pdb.close()
+        compound_pdb.close()
```

Comment:
I think this logic will only close the _last_ compound in the list of compounds stored in the `compound` argument.

Reply:
That is a good catch since I am only closing one file. However, I am now curious how Python handles it. If I check `tmp` before and after running the `mwe.py` from #469 the amount of files are the same.

Maybe that `finally` clause also cleans up the other files?

I guess the cleanest way to make sure all of these files are closed would be to append all these file objects to a list and then iterate through and delete them afterwards.

Reply:
That result doesn't seem intuitive to me, isn't a new `compound_pdb` object being generated each iteration?

This also means a unit test, if we can manage one, would need to have a list of compounds instead of a single one.